### ISSUE TITLE
Updates should always download the new files

### ISSUE THREAD

**Issue by bradbulger** [backdrop/backdrop-issues#5176]
**Description of the bug**

When you update a module or theme to a more recent version using the admin UI, at the point when it would download the appropriate zip file, it checks to see if the file already exists. That check runs through installer_delete_file_if_stale(), which won't remove any previously downloaded copy if it is less than BACKDROP_MAXIMUM_TEMP_FILE_AGE old (6 hours by default).

The result can be that it will "update" the module or theme with the old previously downloaded copy, and no actual update takes place, with no visible indication of what has happened. Or not happened, as it may be.

**Steps To Reproduce**

To reproduce the behavior:
1. Install an older version of a module through the admin UI (use the 'Change release' link)
2. Try to update that module to the current version
3. After the process finishes, check to see if the module has been updated.

**Actual behavior**

The module will still be the version you installed.

**Expected behavior**

The module will have been upgraded to the current version.

**Comment by olafgrabienski** [comment]
Confirmed! I just installed an own contrib theme on a site, made a new release, and tried to update the theme via the user interface. Didn't work (as described above).

**Comment by indigoxela** [comment]
I know that problem.

The trick is to delete the temporary download cache, for instance /tmp/update-cache-369601d7

The file in there is still the old one, the `BACKDROP_MAXIMUM_TEMP_FILE_AGE` hasn't passed yet, so Backdrop won't actually download the new file, but use that (recently) downloaded file.

Responsible for cleaning up that directory is function installer_delete_file_if_stale(), but that doesn't do any intelligent check of the actual version of that module / theme / layout. The filename of that file doesn't ship with any information of the included version - it's simply blaproject.zip. Getting the version requires extracting the file.

This leads to a confusing situation, where the update form claims to download the new file (the information regarding available versions is up-to-date and correct), but function installer_manager_file_get() just takes the old zip file from the download cache directory.

**Comment by bradbulger** [comment]
yes. that is literally in the description of the bug.

**Comment by bradbulger** [comment]
yes. that is what is in the bug description.

i am saying, should it work like that? shouldn't the updater always do the update it claims to be doing? vs trying to be clever with caching and causing this problem?

i don't see the advantage of silently failing.

**Comment by indigoxela** [comment]
> yes. that is what is in the bug description.

Sorry for confirming. :stuck_out_tongue_winking_eye: 

> should it work like that?

I can't answer that. I already tried to explain, why it might be hard to do that in a smarter way, and downloading the same file again and again seems silly, too.

Maybe the question here is: what is the more likely situation? That somebody downloads an outdated version and shortly after that the recent one? That somebody downloads the same version - for which caching makes sense?

Both seem uncommon to me. Both sort of edge cases, unless for developy stuff.

**Comment by kswan** [comment]
I can confirm that this issue still occurs. I installed a module and reported a bug. The module developer quickly fixed the bug and released a new version. The new version was visible in b.org and github, but every attempt to install it didn't apply the update.

Also, a tip regarding the workaround of deleting the update-cache, Ubuntu 20.04 and probably other distros have the update-cache in a systemd directory. In my case it was /tmp/systemd-private-75238f36a9a9b8d3-apache2.service-6oxgVh/tmp/update-cache-96d85067/

**Comment by kswan** [comment]
Thinking about this some more. If b.org stored a hash of each release, the installer could compare hashes to determine whether a fresh download is necessary. 
Currently, the installer looks at the cache directory for a file with the module name (no version number is easily available).
https://github.com/backdrop/backdrop/blob/4841357e0f728aaf882127282faac2873c15c85c/core/modules/installer/installer.manager.inc#L1066

Another perspective to consider is the question of how many duplicate downloads are avoided by the 6 hour cache? Is the 6 hour cache really a benefit?

**Comment by argiepiano** [comment]
> Another perspective to consider is the question of how many duplicate downloads are avoided by the 6 hour cache? Is the 6 hour cache really a benefit?

Agreed. Using a 6 hour cache seems unnecessary in this particular case.

**Comment by klonos** [comment]
Just noting that the default time for garbage collection (the service that runs during cron in Drupal and deletes files if they are "stale", which means that they are being used in 0 places) is 6 hours. There must be some historic reason why that time is set to 6hours, but I also don't know what that reason is - perhaps an arbitrary amount of time that was chosen many years ago 🤷🏼‍♂️

**Comment by indigoxela** [comment]
It seems to me, we agree, that 6 hours of caching doesn't make much sense.

But should we drop the caching completely (as the issue title suggests), or should we just reduce the time. If so, how much time would make more sense?
